Adds SetJsonValue and more documentation

PimDeWitte opened the pull request:
Put it in a new PR for you to make it easier to spot diffs, and so that we always go master->master and not feature branch->master

Commit:
Added SetJson function so that we are not limited to MiniJSON serialization

Commit:
Removed comment, self explanatory

Commit:
Further documentation

Commit:
Further documentation

Commit:
Further documentation

Commit:
Firebase SetJsonValue does not work, this will help reproduce

Commit:
Reproduction

Commit:
Fix for SetJsonValue

Commit:
Fix for SetJsonValue android

Commit:
Updated unity package

Commit:
Reverts test script to not include get wrecked

Commit:
Updates unity package with updated test script

Commit:
Merge pull request #1 from PimDeWitte/feature-enhancements

Feature enhancements

Commit:
Fixes syntax for android

Commit:
Fixes syntax for android in unity package

Commit:
Merge pull request #2 from PimDeWitte/feature-enhancements

Feature enhancements

benwulfe commented on README.md:
```diff
@@ -61,6 +63,77 @@ You will find the API is very similar to the Java/iOS versions except that:
+
+  public class ExampleMainThreadQueue : MonoBehaviour {
+
+    public readonly static Queue<Action> ExecuteOnMainThread = new Queue<Action>();
```
is this threadsafe?  Maybe use concurrentQueue.

PimDeWitte replied:
I think the Enqueue() is not thread-safe. Jon Skeet answered it for us.. :-) http://stackoverflow.com/questions/3870713/is-this-use-of-a-static-queue-thread-safe

If we would call the Enqueue() function inside a coroutine it would be okay, but that would be pretty ugly code. 

using System.Collections.Concurrent is not available in Unity. (http://answers.unity3d.com/questions/59446/is-systemcollectionsconcurrent-or-equivalent-suppo.html)

So to avoid making this too complicated on people, I'm just going to write quick object locking and access control class and link to it as a gist, will add to next PR. 

Nice catch ben!

benwulfe commented on empty-project/Assets/TestScript.cs:
```diff
@@ -36,7 +36,8 @@ public class TestScript : MonoBehaviour {
 		};
 
-		//firebase.SetValue ("working?");
+		firebase.SetValue ("SetValue working?");
```
so there was an issue with setJSON -- was that fixed?  Is this PR ready to go?

PimDeWitte replied:
There was an issue on reading Serialized JSON as a string. You have to make
sure to use the Datasnapshot.DictionaryValue instead of String Value
otherwise it causes that ios issue you helped me fix. That's all. Works
like a charm for the rest.
On Thu, Mar 31, 2016 at 6:13 PM benwulfe notifications@github.com wrote:

> In empty-project/Assets/TestScript.cs
> https://github.com/firebase/Firebase-Unity/pull/29#discussion_r58136662:
> 
> > @@ -36,7 +36,8 @@ public class TestScript : MonoBehaviour {
> >             Debug.Log ("Child removed!");
> >         };
> > -       //firebase.SetValue ("working?");
> > -       firebase.SetValue ("SetValue working?");
> 
> so there was an issue with setJSON -- was that fixed? Is this PR ready to
> go?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/firebase/Firebase-Unity/pull/29/files/b809144708513a7afd3f0a967a9bcd4e8bf3f603#r58136662

Commit:
Add fix for executing functions from the main thread, now thread-safe

Commit:
Merge branch 'master' of github.com:PimDeWitte/Firebase-Unity into feature-enhancements

Commit:
Update library to class

Commit:
Merge pull request #3 from PimDeWitte/feature-enhancements

Feature enhancements

PimDeWitte commented:
Ok @benwulfe - Made a thread safe implementation with documentation and added it to README.md. If you'd like we can also add it to the Unity Firebase SDK, up to you. I just added it like this for now (On my own unofficial branch) because I wasn't sure if the code is up to Google standards. Library and docs on the UnityMainThreadDispatcher are here: 
https://github.com/PimDeWitte/UnityMainThreadDispatcher

Commit:
Fixes ugly indention

Commit:
Merge pull request #4 from PimDeWitte/feature-enhancements

Fixes ugly indention

PimDeWitte commented:
Hi ben,

This one is ready to merge and we're using it. You can merge whenever, PS, if you want to see the finished product: http://pimdewitte.me/2016/04/04/event-driven-matchmaking-with-firebase-unity/

Cheers

PimDeWitte commented:
@benwulfe anything on this? :-)
